linera-sdk/src/lib.rs

```diff
@@ -265,71 +279,84 @@ pub struct QueryContext {
 
 /// Externally visible results of an execution. These results are meant in the context of
 /// the application that created them.
-#[derive(Debug, Default, Deserialize, Serialize)]
+#[derive(Debug, Deserialize, Serialize)]
 #[cfg_attr(any(test, feature = "test"), derive(Eq, PartialEq))]
-pub struct ExecutionResult {
+pub struct ExecutionResult<Effect> {
     /// Sends messages to the given destinations, possibly forwarding the authenticated
     /// signer.
-    pub effects: Vec<(Destination, bool, Vec<u8>)>,
+    pub effects: Vec<(Destination, bool, Effect)>,
     /// Subscribe chains to channels.
     pub subscribe: Vec<(ChannelName, ChainId)>,
     /// Unsubscribe chains to channels.
     pub unsubscribe: Vec<(ChannelName, ChainId)>,
 }
 
-impl ExecutionResult {
+impl<Effect> Default for ExecutionResult<Effect> {
+    fn default() -> Self {
+        Self {
+            effects: vec![],
+            subscribe: vec![],
+            unsubscribe: vec![],
+        }
+    }
+}
+
+impl<Effect: Serialize + std::fmt::Debug + DeserializeOwned> ExecutionResult<Effect> {
     /// Adds an effect to the execution result.
-    pub fn with_effect(
-        mut self,
-        destination: impl Into<Destination>,
-        effect: &impl Serialize,
-    ) -> Self {
-        let effect_bytes = bcs::to_bytes(effect).expect("Effect should be serializable");
-        self.effects.push((destination.into(), false, effect_bytes));
+    pub fn with_effect(mut self, destination: impl Into<Destination>, effect: Effect) -> Self {
+        self.effects.push((destination.into(), false, effect));
         self
     }
 
     /// Adds an authenticated effect to the execution result.
     pub fn with_authenticated_effect(
         mut self,
         destination: impl Into<Destination>,
-        effect: &impl Serialize,
+        effect: Effect,
     ) -> Self {
-        let effect_bytes = bcs::to_bytes(effect).expect("Effect should be serializable");
-        self.effects.push((destination.into(), true, effect_bytes));
+        self.effects.push((destination.into(), true, effect));
         self
     }
 }
 
 /// The result of calling into a user application.
-#[derive(Debug, Default, Deserialize, Serialize)]
+#[derive(Debug, Deserialize, Serialize)]
 #[cfg_attr(any(test, feature = "test"), derive(Eq, PartialEq))]
-pub struct ApplicationCallResult {
+pub struct ApplicationCallResult<Effect, Value, SessionType> {
     /// The return value.
-    #[debug(with = "linera_base::hex_debug")]
-    pub value: Vec<u8>,
+    // #[debug(with = "linera_base::hex_debug")]
+    pub value: Option<Value>,
```

Comment:
I think we should use `Value` here and let the authors of each application decide if they want `Self::ApplicationCallReturnValue` to implement `Default` in order to benefit from the `impl Default` below.